File: 백준/Gold/16236. 아기 상어/아기 상어.py
dx = [1, -1, 0, 0]  # 아 위 오 왼
dy = [0, 0, 1, -1]
import heapq


def is_inside(x, y, N, M):
    return 0 <= x < N and 0 <= y < M


import copy
from collections import *
import sys


# sys.setrecursionlimit(100000)
def input(): return sys.stdin.readline().rstrip()


N = int(input())
a = [list(map(int, input().split())) for _ in range(N)]
X, Y, P = 0, 0, 2
v = []
for i in range(N):
    for j in range(N):
        if a[i][j] == 9:
            X, Y = i, j
            a[i][j]=0
        elif a[i][j]:
            v.append((987654321, i, j,a[i][j]))


def swim(sx, sy, ex, ey, p):
    q = deque()
    q.append((0, sx, sy))
    visited = [[0 for _ in range(N + 1)] for _ in range(N + 1)]
    visited[sx][sy] = 1
    while q:
        cost, x, y = q.popleft()
        if (x, y) == (ex, ey):
            return cost
        for i in range(4):
            n_x = x + dx[i]
            n_y = y + dy[i]
            if is_inside(n_x, n_y, N, N) and not visited[n_x][n_y] and p >= a[n_x][n_y]:
                visited[n_x][n_y] = 1
                q.append((cost + 1, n_x, n_y))
    return 987654321


def chk(sx, sy, p):
    v2 = []
    for cur in v:
        cost, ex, ey,size = cur[0], cur[1], cur[2],cur[3]
        if not a[ex][ey]: continue
        if p > size:
            n_cost = swim(sx, sy, ex, ey, p)
            v2.append((n_cost, ex, ey,size))
        else:
            v2.append((987654321, ex, ey,size))
    v2.sort(key=lambda x: (x[0], x[1], x[2]))
    return v2


cnt = 0
res = 0
while True:
    v = chk(X, Y, P)
    if v and v[0][0] != 987654321:
        X, Y = v[0][1], v[0][2]
        a[X][Y] = 0
        cnt += 1
        res += v[0][0]
        if cnt == P:
            P += 1
            cnt = 0
    else:
        break

print(res)

File: 백준/Gold/16236. 아기 상어/아기 상어.cc
#include <bits/stdc++.h>
#include<unordered_map>
#define MAX 987654321
#define MIN -987654321
#define MOD 1000000009
#define pb push_back
#define pob pop_back
#define ms memset
#define FOR(i,a,b) for(int i = a; i < b; i++)
using namespace std;
typedef long long int ll;
typedef pair<int, int> pi;
typedef pair<int, string> pi3;
typedef pair<int, pair<int, int>>pi2;
typedef vector<vector<int>> vvec;
typedef vector<int> vec;
typedef vector<pi> vec2;
typedef queue<pi> que;
typedef tuple<int, int, int> tup;
typedef vector<vector<pi>> gra;
int dx[4] = { 0,0,-1,1 };
int dy[4] = { 1,-1,0,0 };
int N, M;
int T;
int K;
int res;
int num, num2;
string str;
queue<tup>q;
int m[21][21];
int visited[21][21];
deque<tup>fish;
bool is_inside(int x,int y)
{
	return x >= 0 && x < N&& y >= 0 && y < N;
}
int find_possible(int x, int y,int X,int Y,int cost)
{
	queue<tup>q;
	q.push({ 0,x,y });
	while (!q.empty())
	{
		int cnt, cur_x, cur_y;
		tie(cnt, cur_x, cur_y) = q.front();
		q.pop();
		if (cur_x == X && cur_y == Y)
		{
			return cnt;
		}
		for (int i = 0; i < 4; i++)
		{
			int n_x = cur_x + dx[i];
			int n_y = cur_y + dy[i];
			if (visited[n_x][n_y] || cost < m[n_x][n_y] || !is_inside(n_x, n_y))continue;
			visited[n_x][n_y] = 1;
			q.push({cnt+1,n_x,n_y });
		}
	}
	return 0;
}
int solve()
{
	int cost, x, y;
	tie(cost, x, y) = q.front();
	q.pop();
	int eat_cnt = 0;
	while (1)
	{
		int i;
		int cnt = 0;
		int flag = 0;
		vector<pi>temp;
		int dist = MAX;
		int idx = -1;
		int n_x, n_y;
		for (i = 0; i < fish.size(); i++)
		{
			int f_cost, f_x, f_y;
			tie(f_cost, f_x, f_y) = fish[i];
			if (f_cost >=cost)break;
			memset(visited, 0, sizeof(visited));
			visited[x][y] = 1;
			int next_dist = find_possible(x, y, f_x, f_y, cost);
			if (next_dist)
			{
				if (dist >= next_dist)
				{
					if (dist == next_dist)
					{
						if (n_x >=f_x)
						{
							if (n_x == f_x)
							{
								if (n_y > f_y)
								{
									n_x = f_x;
									n_y = f_y;
									idx = i;
								}
							}
							else
							{
								n_x = f_x;
								n_y = f_y;
								idx = i;
							}
						}
					}
					else if (dist > next_dist)
					{
						n_x = f_x;
						n_y = f_y;
						dist = next_dist;
						idx = i;
					}
				}
			}
		}
		if (idx == -1)
			break;
		else
		{
			x = n_x;
			y = n_y;
			eat_cnt++;
			if (eat_cnt == cost)
			{
				cost++;
				eat_cnt = 0;
			}
			m[n_x][n_y] = 0;
			fish.erase(fish.begin() + idx);
			res += dist;
		}
		if (fish.empty())return res;
	}
	return res;
}
int main()
{
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);
	int i, j;
	cin >> N;
	for (i = 0; i < N; i++)
	{
		FOR(j, 0, N)
		{
			cin >> m[i][j];
			if (m[i][j])
			{
				if (m[i][j] == 9)
				{
					q.push({ 2, i,j });
					m[i][j] = 0;
				}
				else
					fish.push_back({ m[i][j], i, j });
			}
		}
	}
	sort(fish.begin(), fish.end());
	cout<<solve();
}





